### PR TITLE
tools/cephfs: Fix first-damage to include missing dirs

### PR DESCRIPTION
**Pull request by kotreshhr**
Don't skip root and lost+found directories while
checking for damaged metadata.

https://tracker.ceph.com/issues/64615

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/cephfs/first-damage.py [hunk 1]
```diff
@@ -60,14 +60,16 @@
...
+LOST_FOUND_INODE  = "4.00000000"
 
 DIR_PATTERN = re.compile(r'[0-9a-fA-F]{8,}\.[0-9a-fA-F]+')
 
 CACHE = set()
 
 def traverse(MEMO, ioctx):
     for o in ioctx.list_objects():
-        if not DIR_PATTERN.fullmatch(o.key):
+        if not DIR_PATTERN.fullmatch(o.key) and o.key not in [ROOT_INODE, LOST_FOUND_INODE]:
```
Comment (dparmar18) [inline src/tools/cephfs/first-damage.py:72]: I'd say let's use tuple instead of list here

Comment (vshankar) [reply src/tools/cephfs/first-damage.py:72]: why?

Comment (dparmar18) [reply src/tools/cephfs/first-damage.py:72]: no updates to the list as such, only purpose is reading, tuple is better performant, memory efficient than lists.

Comment (vshankar) [reply src/tools/cephfs/first-damage.py:72]: > tuple is better performant, memory efficient than lists.

for just a couple of items?

**Review by batrick — APPROVED** [review]

**Review by vshankar — CHANGES_REQUESTED** [review]
@kotreshhr please add test for this change.

**Comment by kotreshhr** [comment]
> @kotreshhr please add test for this change.

@vshankar I added the test but noticed that it's not part of the test suite and I think that's intentional as per the commit message of e9730461a20739d2368a3473ede5ae4a7b0be2ff

**Comment by vshankar** [comment]
> > @kotreshhr please add test for this change.
> 
> @vshankar I added the test but noticed that it's not part of the test suite and I think that's intentional as per the commit message of [e973046](https://github.com/ceph/ceph/commit/e9730461a20739d2368a3473ede5ae4a7b0be2ff)

Yeh, we don't run in via fs suite. Just share the local execution logs here and I'll be happy to merge :)

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by kotreshhr** [comment]
@vshankar Test results of local run

[result.txt](https://github.com/ceph/ceph/files/14584114/result.txt)

**Comment by kotreshhr** [comment]
[result.txt](https://github.com/ceph/ceph/files/14584129/result.txt)

**Comment by kotreshhr** [comment]
> [result.txt](https://github.com/ceph/ceph/files/14584129/result.txt)

@vshankar Test results which is run locally on vstart cluster.

The snippet where `lost+found` damaged dentry is confirmed that it's removed by first-damage tool
```
+ sudo bin/ceph-fuse -c ./ceph.conf /mnt1
2024-03-13T12:59:16.619+0530 7f460f8d64c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-03-13T12:59:16.623+0530 7f460f8d64c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-03-13T12:59:16.625+0530 7f460f8d64c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-03-13T12:59:16.626+0530 7f460f8d64c0 -1 init, newargv = 0x5648e0641b20 newargc=15
ceph-fuse[36091]: starting ceph client
ceph-fuse[36091]: starting fuse
+ df -h /mnt1
Filesystem      Size  Used Avail Use% Mounted on
ceph-fuse        99G     0   99G   0% /mnt1
+ pushd /mnt1
/mnt1 ~/sandbox/upstream/kotresh-ceph2/ceph/build
+ sleep 5
+ check
+ stat lost+found
stat: cannot statx 'lost+found': No such file or directory
+ popd
~/sandbox/upstream/kotresh-ceph2/ceph/build
+ sudo umount -f /mnt1
```

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
Nice work!

**Pull request merged**